chapter/typeReconstruction.tex

```diff
@@ -85,6 +85,39 @@ \section{Imperative Representation of Terms}
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 \section{ML-Polymorphism}
 
+\newcommand\Let[3]{\texttt{let}\;#1 = #2\;\texttt{in}\;#3}
+\newcommand\FUVars[1]{\mathrm{fuv}(#1)}
+
+
+\begin{alignat*}{2}
+  \tau   & \Coloneqq \alpha \mid \tau \to \tau \mid \UVarX \mid \dots \\
+  \sigma & \Coloneqq \forall\alpha\ldotp\sigma \mid \tau
+\end{alignat*}
+
+\[
+  \{\alpha_1, \dots, \alpha_n\} = \FVars\tau \setminus \FVars\Gamma
+\]
+
+\begin{mathpar}
+  \inferrule{\Gamma \vdash e_1 \colon \tau \\
+               \Gamma, x \colon \forall\alpha_1, \dots, \alpha_n \ldotp \tau
+                 \vdash e_2 \colon \tau'}
+            {\Gamma \vdash \Let{x}{e_1}{e_2} \colon \tau'}
+
+\end{mathpar}
+
+\begin{mathpar}
+  \inferrule{\Gamma \vdash e_1 \Uparrow \tau/\theta \\
+               \theta^*\Gamma, x\colon\forall\alpha_1, \dots, \alpha_n \ldotp
+                 \tau \{ \alpha_1/\UVarX_1, \dots, \alpha_n/\UVarX_n\}
+                 \vdash e_2 \Uparrow \tau'/\theta'}
+            {\Gamma \vdash \Let{x}{e_1}{e_2} \Uparrow \tau'/\theta'\circ\theta}
```

Comment:
I'm not completely sure that the type of x is correct.

Reply:
I looks correct. We substitute α₁ … αₙ for ?X₁ … ?Xₙ and generalize them. Since they only occur in τ we do not have to remember this substitution in returned θ. This is nontrivial observation: they do not occur in Γ, so there is no way that they can escape to other part of typing derivation.